### DISCUSSION TITLE
Accesing Prisma methods by passing in a string as parameter

### DISCUSSION THREAD

**Discussion by anandgeorge** [prisma/prisma#10580] [Q&A]
I have a web form that allows upload of data for a specific table. The form passes in a file and a database table name. How do I create records using the create method passing in the table name as a string in the call to prisma create like so

prisma[table_name].create

Thanks

**Comment by austincrim** [comment]
Hey @anandgeorge 👋!

Using the bracket syntax you mentioned above is exactly how you would use a dynamic property. 

```ts
const tableName = "user"
await prisma[tableName].findFirst() // prisma.user.findFirst()
```

Depending on your exact situation, you might have to add some extra code to satisfy TypeScript.

Let me know if that works!

**Reply by anandgeorge** [reply]
Thanks @austincrim. Could you help me with the code to satisfy Typescript?

**Reply by austincrim** [reply]
Are you running into any specific errors? I'm not sure what your exact setup looks like.

**Reply by anandgeorge** [reply]
It's a typescript project. The editor marks it as an error and typescript won't compile.

**Reply by austincrim** [reply]
Can you provide me with the relevant code snippet or error message?

**Reply by anandgeorge** [reply]
Here's a snapshot of the code as it appears in the editor
![Error](https://user-images.githubusercontent.com/1150098/145137625-e6557ced-90ff-4aef-bd21-0af6d8242c23.png)

```
server.events.on('response', async function ({ info: { id: requestId } }) {
        const files = parseQueue[requestId];

        server.log('start-parsing', {
            files,
            requestId,
            message: 'Start parsing',
        });


        files && files[0] ? fs.createReadStream(files[0])
            .pipe(csv()).on('data', async function(data) {
                console.log(data)
                try {
                    const table = files[1]
                    await prisma[table].create({
                        data
                    })
                } catch (err) {
                    console.log(err)
                }
            })
            .on('end', () => {
                console.log('Finished upload of data')
            })
        : false

    });
```

**Reply by anandgeorge** [reply]
Didn't find the time to look into it. Maybe later.

**Reply by anandgeorge** [reply]
Thanks. Just to be clear I need to pass in a variable table of type sting as prisma[table]. Would this implementation help? Moreover is it good practise to just set the return values of those functions to any? From what I've read in general about modifying types we probably need to go to the index.d.ts files and edit them manually https://stackoverflow.com/questions/40322788/how-to-overwrite-incorrect-typescript-type-definition-installed-via-types-packa

**Reply by saanny** [reply]
can you expose more about your repository ?

**Reply by srevinU** [reply]
@austincrim, your solution works if the variable is declared right before. But what if it is done in a class ? I'm facing a type issue.
There is a sample code:
`export default class Repository {
    
    public model: string = 'event';
    private prisma = new PrismaClient();

    public async getAll(): Promise<Object> {
      const result = await this.prisma[this.model].findMany();
        return result;
      }
      }`
      
And there the error:
Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'PrismaClient<PrismaClientOptions, never, RejectOnNotFound | RejectPerOperation | undefined>'.
     
 Any idea, how would it be manageable ?

**Reply by jharrell** [reply]
Hi @srevinU, 

I believe what is happening is that `model` is being interpreted as a mutable `string` rather than a literal. Could you try replacing your implementation with the following?

```ts
export default class Repository {
  readonly model = 'event'; // note the change to `readonly`
  private prisma = new PrismaClient();

  public async getAll(): Promise<Object> {
    const result = await this.prisma[this.model].findMany();
    return result;
  }
}
```

If it's still not working, please open an issue [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind/bug&projects=&template=bug_report.yml)

**Comment by anandgeorge** [comment]
prisma[table] refers to a key of the the object prisma passed in as a variable. Using a dot notation like prisma.table would imply a key table of the object prisma, not the string that the variable table points to.

**Reply by anandgeorge** [reply]
Not really. It's very common in Javascript to use this notation to refer to keys on objects.

When the PrismaClient sees prisma.user it has to validate that the model User exists. In this case our problem would've been solved if we could wrap the string as an instance of the model. It will then pass vaildation.

Just assume we had a Model class that allows instantiating a model passing in a string. The we could define instantiate

const table = new Model("user")

and then call it using prisma.table.create etc.

My understanding is that you cannot instantiate a model outside of the schema.prisma file. So this approach doesn't work.

**Comment by tit1** [comment]
Im in a simular boat. Im trying to dry up my code and one way was to create a batch update function for Prisma

```javascript
static async batchUpdate(data, table, key){
        for( let d of data){
            let id = d[key]
            delete d[key]
            
            const test = await prisma[table].update({
                where: {
                    [key]: id
                },
                data: d
            })
        }
        return true
    }
```

But this does not work as described above.

**Reply by austincrim** [reply]
Hey there,

Are you experiencing a specific error that you can share?

**Reply by davisgtaylor** [reply]
Did you find a solution to this? Similarly trying to pass in a parameter as a key such that the same prisma call can be used for selections that may have different keys, but otherwise follows the exact same logic. In the example below, the parameter selected will have the new value pushed to it. Any help would be greatly appreciated. 

For example, 


Const parameter = req.body.parameter 
Const value = req.body.value 

Let responses = await prisma.questionnaire.update({
   Where: : {
     Id},
   Data: {
     [parameter] : {
        ...saved[parameter], [value] }
   }
})

**Comment by jmarbutt** [comment]
I am also trying to do something similar, did you ever have any luck?

**Comment by Yohannfra** [comment]
I couldn't find a way around this error. I ended up doing a if/elseif forest like this.

```typescript
    if (ressourceType === 'alentourPoi') {
      ressourceHotelId = (await prisma.alentourPoi.findUnique(/* data */))
    } else if (ressourceType === 'conversation') {
      ressourceHotelId = (await prisma.conversation.findUnique(/* data */))
    } else if (ressourceType === 'faqQuestion') {
    // ....
```

It's ugly but will work for now.

**Comment by HarielThums** [comment]
It is also possible to use it with TS;

```
  async findMany(model: Prisma.ModelName) {
    return this.prisma[model].findMany()
  }
```

**Reply by rhz** [reply]
This made a lot of sense. I tried it, but it doesn't work because `Prisma.ModelName` contains model names in PascalCase while `prisma[model]` expects `model` to be in camelCase. In particular, I get the following error.
```
src/index.ts:320:19 - error TS2551: Property 'UserSignup' does not exist on type 'PrismaClient<PrismaClientOptions, never, DefaultArgs>'. Did you mean 'userSignup'?

320     (await prisma[modelName].findMany({
                      ~~~~~~~~~
```

**Reply by HarielThums** [reply]
@rhz try it

```
  async findMany(model: Uncapitalize<Prisma.ModelName>) {
    return this.prisma[model as Prisma.ModelName].findMany()
  }
```

**Reply by rhz** [reply]
Thank you @HarielThums!
I tried it and it didn't work because of the `model as Prisma.ModelName` with the same error as before. If I use `prisma[model]` instead it gets past that error into a new error. The code looks like
```typescript
  const f = (model: Uncapitalize<Prisma.ModelName>, fieldName: string) =>
    prisma[model].findFirst({
      orderBy: { [fieldName]: 'desc' },
      select: { [fieldName]: true }
    }).then((result) => result?.[fieldName])
```
And the error that it produces is
```
src/index.ts:338:19 - error TS2349: This expression is not callable.
  Each member of the union type '(<T extends UserSignupFindFirstArgs>(args?: SelectSubset<T, UserSignupFindFirstArgs<DefaultArgs>> | undefined) => Prisma__UserSignupClient<GetFindResult<...> | null, null, DefaultArgs>) | ... 6 more ... | (<T extends ItemFindFirstArgs>(args?: SelectSubset<...> | undefined) => Prisma__ItemClient<...>)' has signatures, but none of those signatures are compatible with each other.

338     prisma[model].findFirst({
                      ~~~~~~~~~
```
I wish the Prisma client autogenerated code would include a map from model names to `Prisma.${ModelName}ScalarFieldEnum` so that `fieldName` could be properly typed in the function above. Even then though, I think that wouldn't solve this error.

**Reply by rhz** [reply]
I managed to find a different way to access the field names. The code now looks like
```typescript
  const f = (
    model: Uncapitalize<Prisma.ModelName>,
    fieldName: keyof (Prisma.TypeMap["model"][
      Capitalize<typeof model>]["fields"])) =>
    prisma[model].findFirst({
      orderBy: { [fieldName]: 'desc' },
      select: { [fieldName]: true }
    }).then((result) => result?.[fieldName])
```
It still produces the same error.

**Reply by HarielThums** [reply]
@rhz, try changing `prisma[model]` to `prisma[model as Prisma.ModelName]`, this should work

```
  const f = (model: Uncapitalize<Prisma.ModelName>, fieldName: string) =>
    prisma[model as Prisma.ModelName].findFirst({
      orderBy: { [fieldName]: 'desc' },
      select: { [fieldName]: true }
    }).then((result) => result?.[fieldName])
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!

**Comment by wowjeeez** [comment]
This worked for me in case anyone is still looking:
```ts
type Models = Prisma.TypeMap["model"]


 const count = prismaClient[model as Prisma.ModelName].count as unknown as (input: Models[T]["operations"]["count"]["args"]) => Prisma.PrismaPromise<Models[T]["operations"]["count"]["result"]>
 ```
 
 Heres a utility method:
 ```ts
 export const dynDispatchPrismaCall = <T extends Prisma.ModelName,
    E extends keyof Models[Prisma.ModelName]["operations"],
    A extends Models[T]["operations"][E]["args"]>(model: T, operation: E, input: A): Promise<Models[T]["operations"][E]["result"]> => {
    const prisma = dbCtx()
    const methodRef = prisma[model][operation] as unknown as (input: Models[T]["operations"][E]["args"]) => Prisma.PrismaPromise<Models[T]["operations"][E]["result"]>;
    return methodRef(input)
}
```
